### PR TITLE
Tenancy bug fixes

### LINKED ISSUE #682: Improve the format of "tenant access ls" output
Currently the "tenant access ls" output shows like this:

[root@localhost:/usr/lib/vmware/vmdkops/bin] vmdkops_admin tenant access ls --name MyTenant
Datastore   Create_volume  Delete_volume  Mount_volume  Max_volume_size  Total_size  
----------  -------------  -------------  ------------  ---------------  ----------  
datastore1  1              1              0             128.00MB         1.00GB      

The create/delete/mount privileges should be shown as boolean values ("True" and "False") instead of 1 and 0. Even though the backend db schema type is INTEGER (SQLite doesn't support Boolean), we should be able to convert the output format very easily.

This is not a correctness issue, but I think it's not user friendly and affects the usability a little bit.

### LINKED ISSUE #683: Admin CLI tenant help should print default values
Where applicable the help should print the defaults that will be selected and what do they mean.
Right now I see size as 0 and I am not sure if that means I have no limits or capacity..
```
[root@localhost:~]  /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant access ls --name tenant1
Datastore   Create_volume  Delete_volume  Mount_volume  Max_volume_size  Total_size  
----------  -------------  -------------  ------------  ---------------  ----------  
datastore1  1              1              1             0B               0B   
```

The help does not help in this case 
```
[root@localhost:~]  /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant access add -h
usage: vmdkops_admin.py tenant access add [-h]
                                          [--volume-totalsize Num{MB,GB,TB} - e.g. 2TB]
                                          --name NAME
                                          [--volume-maxsize Num{MB,GB,TB} - e.g. 2TB]
                                          --datastore DATASTORE
                                          [--rights create,delete,mount]

optional arguments:
  -h, --help            show this help message and exit
  --volume-totalsize Num{MB,GB,TB} - e.g. 2TB
                        Maximum total size of all volume that can be created
                        on the datastore for this tenant
  --name NAME           The name of the tenant
  --volume-maxsize Num{MB,GB,TB} - e.g. 2TB
                        Maximum size of the volume that can be created
  --datastore DATASTORE
                        Datastore which access is controlled
  --rights create,delete,mount
                        Datastore access Permissions granted: Choices =
                        ['create', 'delete', 'mount', 'all']
```

### LINKED ISSUE #686: Tenancy: Better error message for docker commands
When a VM is not associated with a VM, the docker commands reports no mount privilege.
```
root@photon-VYb05tsNa [ ~ ]# docker volume ls
list vmdk: VolumeDriver.List: No mount privilege
DRIVER              VOLUME NAME
```

### LINKED ISSUE #688: Tenancy: Tenant rm command needs updates
1. The command I typed below listing volumes to be deleted should not work.
2. The help needs updating.

```
[root@localhost:~] 
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant rm tenant1 --remove-volumes storage3,storage2,storage4
tenant rm succeeded
[root@localhost:~] ls /vmfs/volumes/datastore1/dockvols/tenant1/storage2
storage2-9a71d73b5c60ee6e.vmfd  storage2-flat.vmdk              storage2.vmdk
[root@localhost:~] ls /vmfs/volumes/vsanDatastore/dockvols/tenant1/
.0c558557-03e0-3b0c-6cc4-000c29acb565.lck  storage3-95f889e1f5f5f2b0.vmfd             storage4-c9f772672565851f.vmfd
.17588557-d9db-6d1e-9a44-000c29acb565.lck  storage3.vmdk                              storage4.vmdk
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant ls
Uuid                                  Name     Description  Default_datastore  VM_list  
------------------------------------  -------  -----------  -----------------  -------  
5afa4767-ecd8-489e-9d97-5c46ae50d3fc  DevAppX               default_ds         ubuntu   

[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant rm -h
usage: vmdkops_admin.py tenant rm [-h] [--remove-volumes REMOVE_VOLUMES] name

positional arguments:
  name                  The name of the tenant

optional arguments:
  -h, --help            show this help message and exit
  --remove-volumes REMOVE_VOLUMES
                        BE CAREFUL: Removes this tenant volumes when removing
                        a tenant
[root@localhost:~] 
```

### LINKED ISSUE #703: Multitenancy - Unable to create a volume on a datastore which is not default
Hi Guys,

I have tried the following steps:

1. Create a tenant with VM's. 
vmdkops_admin.py tenant create —name T4 —vm-list photon-new-1

2. Provide access rights for the tenant on a specific datastore. Here the datastore is "nfs-datastore" with all root privileges.
vmdkops_admin.py tenant access add —name T4 —datastore nfs-datastore —rights create,delete,mount

3. Tried to create a volume on the datastore.
docker volume create -d=vmdk —name=Vol4@nfsstore-datastore

Step 3, fails with "Error response from daemon: create Vol4@nfsstore-datastore: VolumeDriver.Create: No create privilege"

But I did give full privileges to the datastore. I can see this on admin CLI.
.vmdkops_admin.py tenant access ls --name T4
Datastore     Create_volume  Delete_volume  Mount_volume  Max_volume_size  Total_size  
------------  -------------  -------------  ------------  ---------------  ----------  
nfsstore-Pop  1              1              1             0B               0B    

When i check the vmdk_ops.log, i can see that its trying to create the VMDK volume on datastore1 which is the default. 

Logs can be seen below.
**11/04/16 19:10:35 447443 [photon-new-1-Tenant100@nfsstore-Pop] [DEBUG  ] get_datastore_name: path=4df30dac-644c9042-b81a-80c16e6cf13a name=['datastore1']**
11/04/16 19:10:35 447443 [photon-new-1-Tenant100@nfsstore-Pop] [DEBUG  ] Authorize: vm_uuid=420b3e7e-0ca1-b647-a78d-92d372e09a7d
11/04/16 19:10:35 447443 [photon-new-1-Tenant100@nfsstore-Pop] [DEBUG  ] Authorize: datastore=datastore1
11/04/16 19:10:35 447443 [photon-new-1-Tenant100@nfsstore-Pop] [DEBUG  ] Authorize: cmd=create
11/04/16 19:10:35 447443 [photon-new-1-Tenant100@nfsstore-Pop] [DEBUG  ] Authorize: opt={u'fstype': u'ext4'}
11/04/16 19:10:35 447443 [photon-new-1-Tenant100@nfsstore-Pop] [DEBUG  ] get tenant vm_uuid=420b3e7e-0ca1-b647-a78d-92d372e09a7d tenant_id=fb612bbb-7c0c-439e-8a2c-299fdd0e4d2f
11/04/16 19:10:35 447443 [photon-new-1-Tenant100@nfsstore-Pop] [DEBUG  ] get_privileges tenant_uuid=fb612bbb-7c0c-439e-8a2c-299fdd0e4d2f datastore=datastore1
11/04/16 19:10:35 447443 [photon-new-1-Tenant100@nfsstore-Pop] [INFO   ] executeRequest 'create' completed with ret={u'Error': 'No create privilege'}

So for whatever datastore except the default datastore I want to create the volume on, it fails with "No create privilege". So, the basic reason is that it always checks for the access rights on default datastore which I don't have in this case.

### PR DESCRIPTION
**Pull request by pdhamdhere**
Fixed following issues;
- Creating volume with default size fails for tenant VM.
- If VM doesn't have access VM Home datastore then `docker volume ls` fails with `list vmdk: VolumeDriver.List: No mount privilege` error. Fixes #686
- If VM doesn't have access to VM Home datastore but have access to nonVMDatastore, user still can't create volumes on nonVMDatastore. Fixes #703 
- Removed auth checks for get/list command.
- "--remove-volume: true" replaced with "—remove-volume"
- "--remove-volume" if specified was deleting all volumes but was not deleting tenant directory.
- Print human readable default values Fixes #683 Fixes #682
- `tenant rm <T_Name>` changed to `tenant rm --name=<T_Name>` Fixes #688 #675

Testing Done:
- `docker volume create -d vmdk --name=vol1` from tenant VM.
- Grant access to nonVMDatastore. Do not grant access to VM Home datastore. 
  - `docker volume ls` succeeds 
  - `docker volume create -d vmdk --name=vol2@nonVMDatastore` succeeds
- For sanity checking, confirmed `docker run` with and without mount privilege.
- Print human readable default values;
```
vmdkops_admin tenant access ls --name=T2
Datastore     Create_volume  Delete_volume  Mount_volume  Max_volume_size  Total_size  
------------  -------------  -------------  ------------  ---------------  ----------  
ssdDatastore  True           True           False         2.00TB           Unset       
```

#668 Unit test should take into account updated behavior in this change.

### CODE REVIEW

File: esx_service/vmdk_ops.py [hunk 1]
```diff
@@ -565,7 +565,7 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
     Returns None (if all OK) or error string
     """
     vm_datastore = get_datastore_name(config_path)
-    error_info, tenant_uuid, tenant_name = auth.authorize(vm_uuid, vm_datastore, cmd, opts)
+    error_info, tenant_uuid, tenant_name = auth.get_tenant(vm_uuid)
```
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Why do we need to call auth.get_tenant here. I think later auth.authorize will call this function inernally?

File: esx_service/utils/auth_data.py [hunk 2]
```diff
@@ -667,12 +666,12 @@ def _remove_volumes_for_tenant(self, tenant_id):
                         return err
```
Comment (lipingxue) [inline esx_service/utils/auth_data.py]: If we hit error when deleting one volume, I think we should continue to delete other volumes. We should return error after go through all volumes.

File: esx_service/vmdk_ops.py (hunk 1 above)
Comment (pdhamdhere) [inline esx_service/vmdk_ops.py]: With this PR we skip authorization checks for list and get command. However, we do need tenantName to list volumes and thus call to get_tenant

File: esx_service/utils/auth_data.py (hunk 2 above)
Comment (pdhamdhere) [inline esx_service/utils/auth_data.py]: Good point. Will fix this.

File: esx_service/cli/vmdkops_admin_test.py [hunk 3]
```diff
@@ -85,7 +85,7 @@ def test_tenant_create_missing_option_fails(self):
         self.assert_parse_error('tenant create')
 
     def test_tenant_rm(self):
-        args = self.parser.parse_args('tenant rm tenant1 --remove-volumes=True'.split())
+        args = self.parser.parse_args('tenant rm tenant1 --remove-volumes'.split())
```
Comment (lipingxue) [inline esx_service/cli/vmdkops_admin_test.py]: It should be 'tenant rm --name tenant1..." since "name" is not a positional argument.

File: docs/user-guide/admin-cli.md [hunk 4]
```diff
@@ -317,22 +317,20 @@ Remove a tenant, optionally all volumes for a tenant can be removed as well.
 
 Sample:
 ```
-/usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant rm tenant1 --remove-volumes True
+/usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant rm --name=tenant1 --remove-volumes
```
Comment (msterin) [inline docs/user-guide/admin-cli.md]: I guess our attempt to follow `docker volume` command backfired. In docker, it is `volume create --name=volname` but `volume rm name`. @kernetime was also commenting on this. So I guess it is confusing. But if we do it consistent and deviate from docker volume command, I'd prefer 'tenant create <name>' and 'tenant rm <name>'

File: esx_service/cli/vmdkops_admin.py [hunk 5]
```diff
@@ -857,8 +859,8 @@ def tenant_rm(args):
     # If args "remove_volumes" is not specified in CLI
     # args.remove_volumes will be None
     if args.remove_volumes:
-        if args.remove_volumes == 'True':
-            remove_volumes = True
+        print "All Volumes will be removed"
```
Comment (msterin) [inline esx_service/cli/vmdkops_admin.py]: this is not Python v3 compatible. Please use function `print("msg")`, not language construct `print msg`

File: docs/user-guide/admin-cli.md (hunk 4 above)
Comment (pdhamdhere) [inline docs/user-guide/admin-cli.md]: Do you mean 
```
vmdkops_admin tenant create T1
vmdkops_admin tenant rm T1
```
Or
```
vmdkops_admin tenant create --name=T1
vmdkops_admin tenant rm --name=T1
```
Most of tenant commands accept tenant name in the form of "--name=T1" for e.g. `vmdkops_admin tenant access ls --name=T1` So, I changed rm to accept  "--name=T1"

File: esx_service/cli/vmdkops_admin.py (hunk 5 above)
Comment (pdhamdhere) [inline esx_service/cli/vmdkops_admin.py]: Done. Looks like all "print" in this file are incompatible with Python 3. I don't want to pollute change so only changed newly added print @ L862.

File: esx_service/cli/vmdkops_admin_test.py (hunk 3 above)
Comment (pdhamdhere) [inline esx_service/cli/vmdkops_admin_test.py]: Yes, looks like I forgot to save file before commit. Will push an update.

File: docs/user-guide/admin-cli.md (hunk 4 above)
Comment (msterin) [inline docs/user-guide/admin-cli.md]: makes sense

File: esx_service/cli/vmdkops_admin.py (hunk 5 above)
Comment (msterin) [inline esx_service/cli/vmdkops_admin.py]: it's OK. I will change others in one pass for P3 move (using 2x3 tool). I just want to avoid new ones popping up

**Comment by govint** [comment]
I do remember fixing this file to be Python 3.x compliant. If adding new
code then can we make it compliant?

On Mon, Nov 7, 2016 at 4:43 AM, Prashant Dhamdhere <notifications@github.com

File: esx_service/vmdk_ops.py [hunk 6]
```diff
@@ -584,6 +584,10 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
                    "Default datastore: %s" \
                    % (datastore, ", ".join(known_datastores()), vm_datastore))
 
+    error_info, tenant_uuid, tenant_name = auth.authorize(vm_uuid, datastore, cmd, opts)
+    if error_info:
+        return err(error_info)
+
```
Comment (govint) [inline esx_service/vmdk_ops.py]: won't this authorize for Get also (called below)

File: docs/user-guide/admin-cli.md (hunk 4 above)
Comment (kerneltime) [inline docs/user-guide/admin-cli.md]: From a usability stand point 
```
vmdkops_admin tenant create T1
vmdkops_admin tenant rm T1
``` 
is a lot nicer, since all the operations are done on a per tenant basis maybe we should change all
```
vmdkops_admin tenant T1 access add --rights all
```

File: esx_service/cli/vmdkops_admin.py (hunk 5 above)
Comment (govint) [inline esx_service/cli/vmdkops_admin.py]: Ok, PR #479 did handle 3.5 compat.

Comment (msterin) [inline esx_service/cli/vmdkops_admin.py]: Yes, but it deviated from the working state (on `print`, and on list iterators, and probably more) so another pass is required. And. more importantly, we need a CI with 6/5, and pylint as a part of the build. And coverage.py in test runs. So for now I will re-fix  on Py3 (or you can do it if you feel like it) and work with @kernetime to add pylinit to the build

File: docs/user-guide/admin-cli.md (hunk 4 above)
Comment (pdhamdhere) [inline docs/user-guide/admin-cli.md]: @kerneltime If we decide to change, let's make it in separate PR.

Comment (msterin) [inline docs/user-guide/admin-cli.md]: @kerneltime - `vmdkops_admin tenant T1 access add --rights all` is not trivial with argparse package. @andrewjstone - can you check if the above is true :-) ?

I suggest leaving it as is

File: esx_service/vmdk_ops.py (hunk 6 above)
Comment (msterin) [inline esx_service/vmdk_ops.py]: @govint - we do not have pylint in the build and no Py3 in CI test bed, so the py3 compatibility can rot. I will add pylint and @kerneltime - can we add py3 (latest ESX GA) to the CI ?

File: docs/user-guide/admin-cli.md (hunk 4 above)
Comment (andrewjstone) [inline docs/user-guide/admin-cli.md]: I'm not really sure what is being asked about here. The way we do choice parsing is by comma seperated values with no spaces. It should be feasible to have `--rights all`  since it's a single value. See [here](https://github.com/vmware/docker-volume-vsphere/blob/master/esx_service/cli/vmdkops_admin.py#L90)
